floors/rule.go

```diff
@@ -0,0 +1,404 @@
+		case Country:
+			value = getDeviceCountry(request)
+		case DeviceType:
+			value = getDeviceType(request)
+		case Channel:
+			value = extractChanelNameFromBidRequestExt(request)
+		case GptSlot:
+			value = getgptslot(imp)
+		case AdUnitCode:
+			value = getAdUnitCode(imp)
+		}
+		ruleKeys = append(ruleKeys, value)
+	}
+	return ruleKeys
+}
+
+func getDeviceType(request *openrtb2.BidRequest) string {
+	value := CATCH_ALL
+	if isMobileDevice(request.Device.UA) {
+		value = Phone
+	} else if isTabletDevice(request.Device.UA) {
+		value = Tablet
+	}
+	return value
+}
+
+func getDeviceCountry(request *openrtb2.BidRequest) string {
+	value := CATCH_ALL
+	if request.Device != nil && request.Device.Geo != nil {
+		value = request.Device.Geo.Country
+	}
+	return value
+}
+
+func getMediaType(imp openrtb2.Imp) string {
+	value := CATCH_ALL
+	if imp.Banner != nil {
+		value = string(openrtb_ext.BidTypeBanner)
+	} else if imp.Video != nil {
+		value = string(openrtb_ext.BidTypeVideo)
+	} else if imp.Audio != nil {
+		value = string(openrtb_ext.BidTypeAudio)
+	} else if imp.Native != nil {
+		value = string(openrtb_ext.BidTypeNative)
+	}
+	return value
+}
+
+func getSizeValue(imp openrtb2.Imp) string {
+	size := CATCH_ALL
+	width := int64(0)
+	height := int64(0)
+	if imp.Banner != nil {
+		if len(imp.Banner.Format) > 0 {
+			width = imp.Banner.Format[0].W
+			height = imp.Banner.Format[0].H
+		} else if imp.Banner.W != nil && imp.Banner.H != nil {
+			width = *imp.Banner.W
+			height = *imp.Banner.H
+		}
+	} else {
+		width = imp.Video.W
+		height = imp.Video.H
+	}
+
+	if width != 0 && height != 0 {
+		size = fmt.Sprintf("%dx%d", width, height)
+	}
+	return size
+}
+
+func getDomain(request *openrtb2.BidRequest) string {
+	value := CATCH_ALL
+	if request.Site != nil {
+		if len(request.Site.Domain) > 0 {
+			value = request.Site.Domain
+		} else {
+			value = request.Site.Publisher.Domain
+		}
+	} else {
+		if len(request.App.Domain) > 0 {
+			value = request.App.Domain
+		} else {
+			value = request.App.Publisher.Domain
+		}
+	}
+	return value
+}
+
+func getSiteDomain(request *openrtb2.BidRequest) string {
+	var value string
+	if request.Site != nil {
+		value = request.Site.Domain
+	} else {
+		value = request.App.Domain
+	}
+	return value
+}
+
+func getPublisherDomain(request *openrtb2.BidRequest) string {
+	var value string
+	if request.Site != nil {
+		value = request.Site.Publisher.Domain
+	} else {
+		value = request.App.Publisher.Domain
+	}
+	return value
+}
+
+func getBundle(request *openrtb2.BidRequest) string {
+	value := CATCH_ALL
+	if request.App != nil {
+		value = request.App.Bundle
+	}
+	return value
+}
+
+func getgptslot(imp openrtb2.Imp) string {
+	var value string
+	adsname, err := jsonparser.GetString(imp.Ext, "data", "adserver", "name")
+	if err == nil && adsname == "gam" {
+		gptSlot, _ := jsonparser.GetString(imp.Ext, "data", "adserver", "adslot")
+		if gptSlot != "" {
+			value = gptSlot
+		}
```

Comment:
What are we supposed to return as the implied else condition here if `gptSlot` is an empty string? Currently we are returning an empty string but is that correct? I see that `getpbsadslot(imp)` returns `"*"` so should we be doing the same here?
I see that most of these functions will default to returning `"*"` which it appears you're doing so these values would at least match the wildcard in the generated rule combinations.

Reply:
If `gptSlot` is an empty string then we are returning` "*"`